File: src/nn-builder/examples/logic.cc
#include <src/nn-builder/src/arch/model.h>
#include <src/nn-builder/src/arch/layers/dense.h>
#include <iostream>
#include <getopt.h>
#include <fstream>

using namespace nn;
using namespace nn::arch;
using namespace nn::arch::layer;
using namespace wabt;

bool FLAG_to_wasm = false;
bool FLAG_to_wat = false;
std::string output_file;

void PrintUsage() {
  std::cout
      << "Logic - Train an OR function" << std::endl
      << "Usage: logic [OPTION]... [FILE]..." << std::endl
      << "    -w, --to-wasm    Print wasm" << std::endl
      << "    -W, --to-wat     Print wat" << std::endl
      << "    -o, --output     Output file" << std::endl
      << "    -h, --help       Display this help message" << std::endl;
}

void InitParams(int argc, char *argv[]) {

  struct option longOptions[] = {

      {"to-wasm", no_argument, 0, 'w'},
      {"to-wat", no_argument, 0, 'W'},
      {"output", required_argument, 0, 'o'},
      {"help", no_argument, 0, 'h'},
      {0, 0, 0, 0}
  };

  int optionIndex = 0;
  int c;
  while ((c = getopt_long(argc, argv, "hwWo:", longOptions, &optionIndex)) != -1) {
    switch (c) {
      case 'w':
        FLAG_to_wasm = true;
        break;
      case 'W':
        FLAG_to_wat = true;
        break;
      case 'o':
        output_file = optarg;
        break;
      case 'h':
      default:
        break;
    }
  }
}

int main(int argc, char *argv[]) {
  InitParams(argc, argv);

  if(!FLAG_to_wat && !FLAG_to_wasm && output_file.empty()) {
    PrintUsage();
    exit(0);
  }

  ModelOptions options;
  options.log_training_accuracy = true;
  options.log_training_error = true;
  options.log_training_time = true;
  options.log_testing_accuracy;
  options.log_testing_error = true;
  options.log_testing_time = true;
  options.log_testing_confusion_matrix = true;
  Model model(options);
  model.SetLayers({
     NewLayer<DenseInputLayer>(2)->WeightType(XavierNormal),
     NewLayer<DenseHiddenLayer>(2, model.Builtins().activation.Sigmoid())->WeightType(XavierNormal),
     NewLayer<DenseOutputLayer>(2, model.Builtins().activation.Sigmoid())->WeightType(LeCunNormal),
  });

  // Train
  std::vector<std::vector<float>> train = {
    {0, 0},
    {0, 1},
    {1, 0},
    {1, 1}
  };
  // Try with one node
  std::vector<std::vector<float>> labels = {
    {1, 0},
    {0, 1},
    {0, 1},
    {0, 1}
  };
  uint32_t epoch = 3000;
  uint32_t training_batch_size = 1;
  uint32_t testing_batch_size = 1;
  uint32_t prediction_batch_size = 4;
  float learning_rate = 0.02;
  auto loss = model.Builtins().loss.MeanSquaredError();
  model.CompileLayers(training_batch_size, testing_batch_size, prediction_batch_size, loss);
  model.CompileTrainingFunction(epoch, learning_rate, train, labels);
  model.CompileTestingFunction(train, labels);
  model.CompilePredictionFunctions();
  model.CompileInitialization();

  assert(model.Validate());
  if(!output_file.empty()) {
    std::ofstream file;
    file.open(output_file);
    if(FLAG_to_wasm) {
      auto data = model.ModuleManager().ToWasm().data;
      file << std::string(data.begin(), data.end());
    } else if(FLAG_to_wat) {
      file << model.ModuleManager().ToWat(true, true);
    }
    file.close();
  }
  return 0;
}

File: src/nn-builder/examples/js/logic.js
if(process.argv.length <= 2) {
  console.log("Missing argument: /PATH/TO/nnb_js.js");
  process.exit(1);
}

process.on('unhandledRejection', error => {
  console.error(">> Make sure SIMD is enabled (e.g. nodejs --experimental-wasm-simd)");
  console.error(">> Error message:", error);
});

const nnb  = require(process.argv[2]);
const {CompiledModel} = require("../../js/compiled_model");

nnb.onRuntimeInitialized = function() {
  // Create model options
  let options = new nnb.ModelOptions();
  options.log_training_error = true;
  options.log_training_accuracy = true;
  options.log_training_time = true;
  options.log_testing_error = true;
  options.log_testing_accuracy = true;
  options.log_testing_confusion_matrix = true;
  options.log_testing_time = true;
  options.log_prediction_results = true;
  options.log_prediction_time = true;
  options.use_simd = true;

  // Create model
  let model = new nnb.Model(options);

  // Create layers
  let l0 = new nnb.DenseInputLayerDescriptor(2);
  l0.SetKeepProb(1.0);
  model.AddDenseInputLayer(l0);

  let l1 = new nnb.DenseHiddenLayerDescriptor(2, "sigmoid");
  l1.SetKeepProb(1.0);
  l1.SetWeightType("xavier_uniform");
  model.AddDenseHiddenLayer(l1);

  let l2 = new nnb.DenseOutputLayerDescriptor(2, "sigmoid");
  l2.SetWeightType("lecun_uniform");
  model.AddDenseOutputLayer(l2);

  // Start compiling
  let data = nnb.ToMatrix([[0, 0], [0, 1], [1, 0], [1, 1]]);
  let labels = nnb.ToMatrix([[1, 0], [0, 1], [0, 1], [0, 1]]);
  let epoch = 10000;
  let training_batch = 1;
  let testing_batch = 1;
  let prediction_batch = 1;
  let loss = "mean-squared-error";
  let learning_rate = 0.02;
  model.SetLayers();
  model.CompileLayers(training_batch, testing_batch, prediction_batch, loss);
  model.CompileTrainingFunction(epoch, learning_rate, data, labels);
  model.CompileTestingFunction(data, labels);
  model.CompilePredictionFunctions();
  model.CompileInitialization();
  if(model.Validate()) {
    let buffer = nnb.ToUint8Array(model.ToWasm());
    let compiled_model = new CompiledModel();
    const lib = WebAssembly.instantiate(buffer, compiled_model.Imports());
    lib.then(wasm => {
      compiled_model.SetWasm(wasm);
      console.log("Training ...");
      compiled_model.Train();
      console.log("Testing ...");
      compiled_model.Test();
      console.log("Predicting ...");
      compiled_model.Predict([[0, 0]]);
      compiled_model.Predict([[0, 1]]);
      compiled_model.Predict([[1, 0]]);
      compiled_model.Predict([[1, 1]]);
    });
  }
}
